filetests/isa/x86/windows_fastcall_x64_unwind.clif

```diff
@@ -147,23 +156,56 @@ block0(v0: i64, v1: i64):
     store.i32 v12, v1+80
     store.i32 v13, v1+88
     store.i32 v14, v1+96
+    store.f64 v15, v1+104
+    store.f64 v16, v1+112
+    store.f64 v17, v1+120
+    store.f64 v18, v1+128
+    store.f64 v19, v1+136
+    store.f64 v20, v1+144
+    store.f64 v21, v1+152
+    store.f64 v22, v1+160
+    store.f64 v23, v1+168
     return
 }
 ; sameln: UnwindInfo {
 ; nextln:     version: 1,
 ; nextln:     flags: 0,
-; nextln:     prologue_size: 19,
-; nextln:     unwind_code_count_raw: 10,
+; nextln:     prologue_size: 42,
+; nextln:     unwind_code_count_raw: 16,
 ; nextln:     frame_register: 5,
 ; nextln:     frame_register_offset: 0,
 ; nextln:     unwind_codes: [
 ; nextln:         UnwindCode {
-; nextln:             offset: 19,
+; nextln:             offset: 42,
 ; nextln:             op: SmallStackAlloc,
-; nextln:             info: 3,
+; nextln:             info: 5,
```

Comment:
Sorry, I just caught this.  I would expect a value of **4** originally, indicating a 40 byte SP adjustment: 8 bytes to realign the stack from the even number of GPR pushes and 32 bytes for the callee spill space we never elide (even for leaf functions).  So there's a bug currently in how we're aligning the stack even when spilling just the GPRs.

With your changes, I would expect a value of 10, indicating 88 bytes: 8 bytes to realign the stack from the even number of GPR pushes, 48 bytes for the saved XMM registers, and 32 bytes for the callee spill space.

Reply:
For comparison, the VC++ example above allocates 152 bytes for its frame: 8 bytes for stack alignment (upon entry stack is always off by 8 bytes due to call instruction pushing return address), 112 bytes for saved XMM registers, and 32 bytes for callee spill space (the function does call).

Reply:
Yep, there's definitely an error here. I hadn't checked the generated assembly on this case but
```
Disassembly of 356 bytes:                                                                                                                                                               [69/1840]
   0:   55                      push    rbp                        
   1:   48 89 e5                mov     rbp, rsp                   
   4:   53                      push    rbx                         
   5:   56                      push    rsi                         
   6:   57                      push    rdi                         
   7:   41 54                   push    r12                         
   9:   41 55                   push    r13                         
   b:   41 56                   push    r14                         
   d:   41 57                   push    r15                         
   f:   48 8d 84 24 10 00 00 00 lea     rax, [rsp + 0x10]           
  17:   f2 0f 11 30             movsd   qword ptr [rax], xmm6       
  1b:   f2 0f 11 78 10          movsd   qword ptr [rax + 0x10], xmm7
  20:   f2 44 0f 11 40 20       movsd   qword ptr [rax + 0x20], xmm8
```
is definitely not the prologue I'd want to see. It looks like the callee-save for fprs begins at where the stack was when created, rather than being a non-overlapping region laid out somewhere later on. Which makes sense, since this is so late in compilation. urgh.

Reply:
Was the unwind information "correct" in that RSP is adjusted by only 0x28 following saving the XMM registers?

Reply:
correct in that it was adjusted by 0x30 (5 * 8 == 40 == 0x28, but there's the implied extra 8 bytes bringing it to 0x30)

Reply:
Oh sorry, I meant 0x30 (the 5 in the unwind info indicating (8*5)+8; I seem to always forget the +8).  Thanks for confirming.

Reply:
Also, judging from the above disassembly, The offsets from RSP for the XMM registers are off by 1 in the resulting unwind information it seems.  We're recording the offsets from RAX in the unwind info (scaled-by-16 values of 0, 1, and 2), but RAX itself is offset 0x10 from RSP.

Reply:
I've deleted my last comment since it's probably just related to the overlapping stack slot issues.  Once we have the proper prologue generated, I'll verify that the unwind information properly describes it.

Reply:
Some notes as it took me a surprising amount of time to figure out what _exactly_ was going on. I didn't really understand `StackSlot` too well.

The gist of it is that `lea rax, [rsp + 0x10]` may look like an address of nothing in particular, but that's because it seems `StackSlot` used alongside adjusting the stack is a recipe for bad times. It's the instruction that would be correct to get the stack slot's start at the entry of the function without intervening changes. That is, the return address and `rbp` are preserved on the stack by Cranelift, and 0x10 down the stack from the base address of the call frame would be `rbx` and the start of the rest of the csr region. `rsp` isn't really "ready for use" until after the prologue though, so `stack_addr` doesn't seem appropriate to use as it currently exists for this purpose.

Subsequently, the xmm stores _would_ overwrite the return address, `rbp`, `rbx`, ..., were the stack slot address right - I misunderstood which direction offsets count from too.

I think what I can do is make a second stack slot that we can directly address for FPR save/lrestore, and requiring 16-byte alignment on that would keep offsets simple as well. I'll give it a try and we'll see how it goes :crossed_fingers:

Reply:
Surprise of surprises, the number is now not `3`, `4`, `5`, or `10`, but `12`! I can argue the correctness of 12, though. For `stack_slot` reasons I don't understand, the FPR preservation region is aligned more strictly than necessary. This grows what would be 48 bytes into 64. Plus eight bytes to realign the stack, and 32 bytes of shadow stack space, that makes for a 104-byte allocation. That would be 13 8-byte slots, minus one for the implied slot makes 12!

A similar wider-than-necessary FPR region is also present in the [`float_callee_save`](https://github.com/bytecodealliance/cranelift/blob/399fb63390f66592730f9922f89b980b16a1ad22/filetests/isa/x86/windows_fastcall_x64.clif#L49) test I added. I've tried to follow through what causes this in `layout_stack` and my best theory is that treating a stack slot as a collection of smaller units is not quite how it's intended to be used, and instead it tries to align the provided stack slots using their sizes as an alignment requirement. It looks like treating a stack slot as an aggregate of more permissible units might be a bit of a change, so I'm open to recommendations if taking ~16-48 extra bytes per call frame is a nonstarter.